src/protocol/command.rs

```diff
@@ -41,130 +42,24 @@ impl Command {
-                        Some(' ') => ' ',
-                        Some('\\') => '\\',
-                        Some(char) => {
-                            self.parse_error(&format!("unknown escaped character {}", char))?
-                        }
-                    }),
-                    Some(char) => {
-                        if excluded.contains(&char) {
-                            self.input.push_front(char);
-                            None
-                        } else {
-                            Some(char)
-                        }
-                    }
-                })
-            }
-
-            fn collect_chars_until(&mut self, excluded: &[char]) -> R<String> {
-                let mut result = "".to_string();
-                while let Some(char) = self.parse_char(excluded)? {
-                    result.push(char);
-                }
-                Ok(result)
-            }
-
-            fn parse_word(&mut self) -> R<Option<String>> {
-                self.skip_spaces();
-                Ok(match self.input.get(0) {
-                    None => None,
-                    Some('"') => {
-                        self.skip_char('"', "shouldn't happen")?;
-                        let word = self.collect_chars_until(&['"'])?;
-                        self.skip_char('"', "unmatched quotes")?;
-                        self.peek_chars(
-                            vec![Some(' '), None],
-                            "closing quotes must be followed by a space",
-                        )?;
-                        Some(word)
-                    }
-                    Some(_) => {
-                        let result = self.collect_chars_until(&[' ', '"'])?;
-                        self.peek_chars(
-                            vec![Some(' '), None],
-                            "opening quotes must be preceeded by a space",
-                        )?;
-                        Some(result)
-                    }
-                })
-            }
-
-            fn parse_command(&mut self) -> R<Command> {
-                let executable = match self.parse_word()? {
-                    None => self.parse_error("expected: space-separated command and arguments")?,
-                    Some(executable) => executable.into_bytes(),
-                };
-                let mut arguments = vec![];
-                loop {
-                    match self.parse_word()? {
-                        None => break,
-                        Some(word) => arguments.push(word.into_bytes()),
-                    }
-                }
-                Ok(Command {
-                    executable,
-                    arguments,
-                })
-            }
-        }
-
-        Ok(Parser {
-            original: string.to_string(),
-            input: string.chars().collect(),
-        }
-        .parse_command()?)
+    pub fn new(command: &str) -> R<Command> {
+        Parser::parse_arguments(command)?
+            .split_first()
```

Comment:
`split_first` is nice. I always wanted a replacement for pattern matching on list cons'es.